client/client.go

```diff
@@ -288,138 +288,111 @@ func (c *client) tsLoop() {
 	loopCtx, loopCancel := context.WithCancel(c.ctx)
 	defer loopCancel()
 
-	streams := make(map[string]pdpb.PD_TsoClient)
+	ticker := time.NewTicker(tsLoopDCCheckInterval)
+	defer ticker.Stop()
 	for {
-		ctx, cancel := context.WithCancel(loopCtx)
-		if !c.checkTSOStreams(streams) {
-			if err := c.createTSOStreams(ctx, streams); err != nil {
-				select {
-				case <-loopCtx.Done():
-					cancel()
-					return
-				default:
-				}
-				c.ScheduleCheckLeader()
-				cancel()
-				c.revokeTSORequest(errors.WithStack(err))
-				select {
-				case <-loopCtx.Done():
-					return
-				case <-time.After(time.Second):
-				}
-				continue
-			}
-		}
-		var err atomic.Value // stored as error
-		requestsDispatcher := make(map[string][]*tsoRequest)
-		// Initialize the requests dispatcher
-		for dcLocation := range streams {
-			requestsDispatcher[dcLocation] = make([]*tsoRequest, 0, maxMergeTSORequests+1)
-		}
-		// Dispatch different dc-locations requests to corresponding TSO stream
-		select {
-		case first := <-c.tsoRequests:
-			requestsDispatcher[first.dcLocation] = append(requestsDispatcher[first.dcLocation], first)
-			for i := 1; i <= len(c.tsoRequests); i++ {
-				req := <-c.tsoRequests
-				requestsDispatcher[req.dcLocation] = append(requestsDispatcher[req.dcLocation], req)
-			}
-			wg := sync.WaitGroup{}
-			for dcLocation, requests := range requestsDispatcher {
-				if len(requests) == 0 {
-					continue
-				}
-				if _, exist := streams[dcLocation]; !exist {
-					err := errs.ErrClientGetTSO.FastGenByArgs(fmt.Sprintf("unknown dc-location %s to the client", dcLocation))
-					log.Error("[pd] dispatch tso request error", zap.String("dc-location", dcLocation), errs.ZapError(err))
-					c.finishTSORequest(requests, 0, 0, err)
-					continue
-				}
-				wg.Add(1)
-				go func(dc string, reqs []*tsoRequest) {
-					defer wg.Done()
-					done := make(chan struct{})
-					dl := deadline{
-						timer:  time.After(c.timeout),
-						done:   done,
-						cancel: cancel,
+		c.allocators.Range(func(dcLocationKey, _ interface{}) bool {
```

Comment:
Does this affect the performance?  before no need mutex.

Reply:
You can check the [source codes](https://golang.org/src/sync/map.go#L308) of the `Range` method. It mostly iterates over a read-only map data, which won't block other writing or reading operations and doesn't involve a `Mutex` in most cases.

Reply:
got it!